File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    de = deque([[0,0,1]])
    move = [[1,0], [-1,0], [0,1], [0,-1]]
    row = len(maps)
    col = len(maps[0])
    flag = 0
    while (de):
        player = de.popleft()

        for i in range(4):
            x = player[0] + move[i][0]
            y = player[1] + move[i][1]
            if (x == (row - 1) and y == (col - 1)):
                return player[2] + 1
            elif (x >= 0 and x < row and y >= 0 and y < col and maps[x][y] != 0):
                de.append([x,y,player[2] + 1])
                maps[x][y] = 0
        
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    let answer = 0;
    const row = maps.length;
    const col = maps[0].length;
    const queue = [[0,0,0]];
    const move = [[0, 1], [0, -1], [1, 0], [-1, 0]];
    maps[0][0] = 0;
    while (queue.length !== 0) {
        const [r, c, cnt] = queue.shift();

        if (r === row - 1 && c === col - 1) {answer = cnt; break;}
        
        for (let i = 0; i < 4; i++) {
            const nr = move[i][0] + r;
            const nc = move[i][1] + c;

            if (nr >= 0 && nr < row && nc >= 0 && nc < col && maps[nr][nc] === 1) {
                maps[nr][nc] = 0;
                queue.push([nr, nc, cnt + 1]);
            }
        }
    }
    if (answer === 0) answer = -2;
    return answer + 1;
}